Add support for UNAssociationsPopupElement to Popup

puneet-pdx opened the pull request:
<!-- PRs should provide sufficient context on the changes, either by linking to the associated issue and/or by providing a summary. Also provide a link to the design if it's appropriate. -->

Related to issue: # https://devtopia.esri.com/runtime/apollo/issues/1407
 
<!-- link to design, if applicable -->

### Description:

Add support for UNAssociations in Popups. 

### Summary of changes:

- Add `PopupState`, Add new Popup Composable that takes the State object
- Update all internal popup elements to not have remember functions and just initialize them using constructors and hold them in the PopupState object
- Add UtilityAssociationElement(from FeatureForms)
- Add support for Navigation to enable browsing from one Popup to another based on an association element(from FeatureForms)
- Add PopupScreen
- Add ContentAwareTopBar(from FeatureForms)
- Deprecate stateful PopupComposable
- Update MicroApp to use the new Composable

### Pre-merge Checklist

<!-- Tick one box for each section -->
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/937/
  - [x] https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/946/
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [ ] Yes
  - [ ] No

puneet-pdx committed:
WIP

puneet-pdx committed:
WIP

puneet-pdx committed:
Update Popup and UtilityAssociationsElementState

puneet-pdx committed:
WIP

puneet-pdx committed:
WIP

puneet-pdx committed:
clean up code

puneet-pdx committed:
clean up imports and unused code

puneet-pdx committed:
WIP

puneet-pdx committed:
remove duplicate title from old popup
remove support for hasEdits from contentAwareTopbar

puneet-pdx committed:
move popupelementstates creation to PopupState
replace form references with popup
remove caching attachments in popupstate

puneet-pdx committed:
update micro app

puneet-pdx committed:
optimize imports

puneet-pdx committed:
enable isNavigationEnabled
add doc

puneet-pdx committed:
add support for ignore list

puneet-pdx committed:
add support for dynamic entity

puneet-pdx committed:
update api file
deprecate popup composable

puneet-pdx committed:
update micro app

puneet-pdx committed:
update micro app to not show close icon

puneet-pdx committed:
add copyright
disable lint for missing translations

puneet-pdx committed:
Merge branch 'v.next' into puneet/6396_UNAssociationsPopup

# Conflicts:
#	toolkit/popup/api/popup.api

kaushikrw commented on microapps/PopupApp/app/src/main/java/com/arcgismaps/toolkit/popupapp/screens/mapscreen/MapViewModel.kt:
```diff
@@ -106,45 +87,8 @@ class MapViewModel(
-            }
-        }
+    fun updatePopupState(popup: Popup?) {
+        _popupState.value = PopupState(popup!!, viewModelScope)
```
This needs a null check for `popup` before you apply the `!!` operator.

puneet-pdx replied:
updated

kaushikrw commented on toolkit/popup/api/popup.api:
```diff
@@ -1,4 +1,69 @@
+	public final fun getPopup ()Lcom/arcgismaps/mapping/popup/Popup;
+}
+
+public final class com/arcgismaps/toolkit/popup/internal/element/fieldselement/FieldsElementState$Creator : android/os/Parcelable$Creator {
```
There are many classes in this api dump that are internal and need to be added to the list of `ignoredClasses` in `build.gradle.kts`.

puneet-pdx committed:
remove parcelize where it is not being used
update API file

puneet-pdx committed:
Merge branch 'v.next' into puneet/6396_UNAssociationsPopup

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/Popup.kt:
```diff
@@ -91,26 +65,63 @@ private data class PopupState(@Stable val popup: Popup)
+ * [UtilityAssociationsPopupElement]s present. When true, the user can navigate to associated features
+ * and back. If false, this navigation is disabled. Default is true
+ *
+ * @since 200.9.0
```
```suggestion
 * @since 300.0.0
```
I don't think there is a 200.9

puneet-pdx replied:
yes, updated

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/Popup.kt:
```diff
@@ -91,26 +65,63 @@ private data class PopupState(@Stable val popup: Popup)
  */
+@Deprecated(
+    message = "Maintained for binary compatibility. Use the overload that uses the PopupState object.",
+    level = DeprecationLevel.HIDDEN
```
Any particular reason this is hidden? Imo this would be confusing to anyone who upgrades to v300 and is not able to see this function. Not sure what the user story for Popup is but imo having this as a warning with later removal or a full removal in 300.0 might be a better approach.

puneet-pdx replied:
I vaguely remember Soren mentioning it(I am not 100%). I have changed it to warning for now. Soren can chime in later.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/Popup.kt:
```diff
@@ -91,26 +65,63 @@ private data class PopupState(@Stable val popup: Popup)
+    }
+
+    if (stateData.getActivePopupStateData().initialEvaluation.value) {
+        Popup(stateData, modifier, showCloseIcon = false)
```
It would be nice to have a progress/loading indicator here until the expressions are evaluated. A simple jump without a transition might feel jarring.

puneet-pdx replied:
This check is not needed here and I have removed it. We are displaying progress/loading indicator in the PopupScreen.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/Popup.kt:
```diff
@@ -91,26 +65,63 @@ private data class PopupState(@Stable val popup: Popup)
+    modifier: Modifier = Modifier,
+    onDismiss: () -> Unit = {},
+    showCloseIcon: Boolean = true,
+    isNavigationEnabled : Boolean = true
```
```suggestion
    isNavigationEnabled : Boolean = true
```
For `FeatureForm`, this was meant as a way for a developer to disable the navigation in case there are geometry edits on the feature (since a form isn't concerned with geometry). Is there a similar need/requirement for a developer to disable navigation for popups?

puneet-pdx replied:
No, I checked the design, there is no requirement. I will remove support for it.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt:
```diff
@@ -0,0 +1,358 @@
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.launch
+
+public class PopupState(@Stable public val popup: Popup) {
```
Doc is missing for the `PopupState` class.

puneet-pdx replied:
added

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt:
```diff
@@ -0,0 +1,358 @@
+     * Indicates if the evaluateExpression function for the [popup] has been run.
+     */
+    internal var initialEvaluation : MutableState<Boolean> = mutableStateOf(false)
+        private set
```
```suggestion
    var initialEvaluation : Boolean by mutableStateOf(false)
        private set
```
This needs to be a var of type `Boolean` for the private setter to work. Otherwise any consumer can simply set a value directly like so -
```kotlin
initialEvaluation.value = false
```

I also don't think there is a need to mark these properties and methods as internal since the class is already marked internal.

Similarly for other properties below too.

puneet-pdx replied:
Updated. We should make similar change for FormStateData as well 

https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/forms/toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureFormState.kt#L331

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt:
```diff
@@ -0,0 +1,358 @@
+
+        this.getFeatureSubtype() != null && this.featureTable is ArcGISFeatureTable -> {
+            (this.featureTable as ArcGISFeatureTable).subtypeSubtables
+                .firstOrNull { it.name == this.getFeatureSubtype()?.name }
```
```suggestion
                .firstOrNull { it.subtype == this.getFeatureSubtype()}
```
The `FeatureSubtype` can be checked directly here.

puneet-pdx replied:
updated

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/fieldselement/FieldsElementState.kt:
```diff
@@ -15,42 +15,30 @@
-        id = PopupElementState.createId()
+        fieldsToFormattedValues = element.fields.map { it.label }
+            .zip(element.formattedValues)
+            .toMap(),
```
```suggestion
        fieldsToFormattedValues = element.fields
            .mapIndexed { index, field -> field.label to element.formattedValues[index] }
            .toMap(),
```

The intermediate collection from `zip` can be avoided by using `mapIndexed` instead.

puneet-pdx replied:
updated

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/media/MediaElementState.kt:
```diff
@@ -55,31 +45,48 @@ import java.util.UUID
-        mediaPopupElement: MediaPopupElement,
+    lateinit var description: String
+    lateinit var title: String
+    lateinit var media: List<PopupMediaState>
```
I'm wondering why these properties are mutable.

puneet-pdx replied:
I have updated them, only media is mutable now as it needs to be created from inside a composable

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/media/MediaElementState.kt:
```diff
@@ -55,31 +45,48 @@ import java.util.UUID
+    internal var isInitialized : MutableState<Boolean> = mutableStateOf(false)
+        private set
+
+    fun init(
```
Similarly, this indicates this `MediaElementState` can be initialized multiple times. Any particular reason why this approach compared to the typical constructor pattern?

puneet-pdx replied:
I have updated it to not be an init function by a create function for creating media

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/media/MediaPopupElement.kt:
```diff
@@ -39,13 +42,27 @@ internal fun MediaPopupElement(
+            media = state.media,
+            onClickedMedia = onClickedMedia
+        )
+    }
```
This initialization approach appears somewhat complex. Since the `init` can be called many times, this makes the state mutable plus the scope provided is associated with the composition, and several refresh or fetch operations depend on it, it might be helpful to consider simplifying this process.

puneet-pdx replied:
I have refactored this a bit, to prevent creation of media on recompositions. Lets discuss...

kaushikrw replied:
This looks good now!

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationDetails.kt:
```diff
@@ -0,0 +1,245 @@
+    val associationResult = state.selectedAssociationResult ?: return
+    val filter = state.selectedFilterResult?.filter ?: return
+    val association = associationResult.association
+    var showConfirmationDialog by remember {
+        mutableStateOf(false)
+    }
```
```suggestion
```
Since Popups are read only the confirmation dialog stuff can be removed.

puneet-pdx replied:
updated

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationGroupResult.kt:
```diff
@@ -0,0 +1,329 @@
+    onClick: () -> Unit,
+    onDetailsClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
```
Some of the UI and rules for this has been updated. It might be a good idea to revisit this before release and update it accordingly.

puneet-pdx replied:
Yes, the code on `v.next` is missing support for viewing details for an association and onDetailsClick is never invoked.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationsElementState.kt:
```diff
@@ -0,0 +1,120 @@
+    override val id : Int = element.hashCode()
+    val label : String = element.title
+    val description: String = element.description
+    //    val isVisible : StateFlow<Boolean> = element.isVisible
```
```suggestion
```

puneet-pdx replied:
updated

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociations.kt:
```diff
@@ -0,0 +1,329 @@
+            modifier = Modifier.clip(shape = RoundedCornerShape(15.dp)),
+            state = lazyListState
+        ) {
+            groupResult.associationResults.forEachIndexed { index, info ->
```
```suggestion
            groupResult.associationResults.forEachIndexed { index, info ->
```
IIRC popup spec defines a [displayCount](https://runtime-zip.esri.com/userContent/runtime/kotlin/doc_kotlin/300.0.0/last_good/html/arcgis-maps-kotlin/com.arcgismaps.mapping.popup/-utility-associations-popup-element/display-count.html) for the number of associations to show. I would check with Swift to see how they've handles this.

puneet-pdx replied:
I have added support for taking in account the display count.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/navigation/NavigationAction.kt:
```diff
@@ -0,0 +1,53 @@
+ * Indicates the action to take when navigating.
+ */
+@Parcelize
+internal sealed class NavigationAction(val value: Int) : Parcelable {
```
I'm not sure Popup needs this `NavigationAction`. For forms, this was meant as a way to cache user actions across configuration changes and dialogs. Since popup is read only a navigation action does not need a confirmation like in forms (which is where they are cached).

puneet-pdx replied:
Ok. I have removed it.

kaushikrw reviewed (changes requested):
@puneet-pdx I wasn't able to load any map to test in the micro-app but here is some initial feedback.

puneet-pdx committed:
address code review comments

puneet-pdx committed:
remove support for isNavigationEnabled and NavigationAction

puneet-pdx committed:
remove showConfirmationDialog in UtilityAssocationsDetails

puneet-pdx committed:
add support for display count
add some suggested optimizations

puneet-pdx committed:
Optimize imports
optimize mediaElementState

puneet-pdx committed:
update logic in mediaPopupElement composable

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/media/MediaElementState.kt:
```diff
@@ -55,31 +47,46 @@ import java.util.UUID
-        scope: CoroutineScope,
+    val description: String = element.description
+    val title: String = element.title
+    lateinit var media: List<PopupMediaState>
```
Any particular reason this is late initialized? `isPopupMediaCreated` should be enough to infer if the media creation is completed, no?

puneet-pdx replied:
Hmm. I see your point. I have changed it to be an empty list on initialization.

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt:
```diff
@@ -0,0 +1,376 @@
+        store.addLast(popupStateData)
+    }
+
+    public constructor(popup: Popup, scope: CoroutineScope) : this(popup) {
```
Missing doc for the public ctor.

puneet-pdx replied:
added

kaushikrw commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/screens/ContentAwareTopBar.kt:
```diff
@@ -0,0 +1,225 @@
+    onBackPressed: () -> Unit,
+    onClose: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
```
One thing I noticed while running the app is, the popup shows up in an expandable group with an empty title. So I was wondering why the expandable group since there is no group element in popups.

puneet-pdx replied:
All the element types are displayed as an expandable cards with title being the title of the element. This is a requirement coming from the design
_A stretch goal is to have the individual view elements be "collapsible". When collapsed, a popup element view just displays the title of the element._
https://devtopia.esri.com/runtime/common-toolkit/blob/main/designs/Popup/Popup.md#miscellaneous

kaushikrw replied:
Ah I see. Perhaps it is a design discussion then, because the collapsed elements do have a title. I will leave that up to you..

kaushikrw reviewed (commented):
Looking good! Just a few more comments below.

puneet-pdx committed:
enable close button
add doc

puneet-pdx commented:
@kaushikrw I have addressed your comments here and also enabled the close button in the micro app. I am looking into updating `UNAssociationsScreen` with the latest changes.

puneet-pdx committed:
add search box to filter UtilityAssociationGroupResults with a user specified string

puneet-pdx committed:
update showall behavior on search
update AssociationItem properties to display with the latest on FeatureBranch

puneet-pdx committed:
Add fix for navigation not working
Change bottomsheet behavior for onDismiss

puneet-pdx committed:
Add default value for UtilityElement title

puneet-pdx committed:
reset geoElement when the bottomsheet is dismissed

kaushikrw commented on microapps/PopupApp/app/src/main/java/com/arcgismaps/toolkit/popupapp/screens/mapscreen/MainScreen.kt:
```diff
@@ -63,13 +68,32 @@ fun MainScreen(viewModel: MapViewModel) {
     )
+    LaunchedEffect(scaffoldState.bottomSheetState.currentValue) {
+        if (scaffoldState.bottomSheetState.currentValue == SheetValue.Hidden) {
+            unselectFeature(viewModel.geoElement, viewModel.layer)
+            viewModel.updatePopupState(null)
+            viewModel.setGeoElement(null)
+        }
+    }
```
These sequence of steps are already happening in the `Popup.onDismiss` event so it seems redundant to do the same here.

puneet-pdx replied:
The `Popup.onDismiss` does not get invoked if the user manually closes the bottomsheet. If that happens then without this there will be no bottomsheet on the screen and there will be a feature selected on the screen. This makes sure that the behavior is consistent for both the workflows.

kaushikrw reviewed (approved):
Looks good! I just have one comment below.

puneet-pdx committed:
Pick up refactor changes from feature forms
Update API file

puneet-pdx committed:
update micro app

sorenoid commented on microapps/PopupApp/app/src/main/java/com/arcgismaps/toolkit/popupapp/screens/mapscreen/MainScreen.kt:
```diff
@@ -63,13 +68,28 @@ fun MainScreen(viewModel: MapViewModel) {
-                Popup(viewModel.popup!!, Modifier.animateContentSize())
+            if (viewModel.popupState != null) {
+                Popup(
+                    viewModel.popupState!!,
```
could use smart casting here rather than the !! operator.

```suggestion
            val state = viewModel.popupState
            if (state != null) {
                Popup(
                    state,
```

sorenoid commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/navigation/PopupNavHost.kt:
```diff
@@ -0,0 +1,150 @@
+                // guard against null value
+                as? UtilityAssociationsElementState ?: return@composable
+            // Display the association details
+            UtilityAssociationDetails(
```
It seems like this function should not be called if the result and filter are null, rather than passing in the state and returning early from a composable function. Could the parameters be 

```kotlin
val result = utilityAssociationsElementState.state.selectedAssociationResult
val filter = state.selectedFilterResult?.filter
if (result != null && filter != null) {
    UtilityAssociationDetails(
        result,
        filter,
        modifier = Modifier.fillMaxSize()
    )
}
```

sorenoid commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationDetails.kt:
```diff
@@ -0,0 +1,183 @@
+    modifier: Modifier = Modifier
+) {
+    val associationResult = state.selectedAssociationResult ?: return
+    val filter = state.selectedFilterResult?.filter ?: return
```
see comment on this function at the call site

sorenoid commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationDetails.kt:
```diff
@@ -0,0 +1,183 @@
+        }
+        associationResult.getFractionAlongEdge()?.let { fraction ->
+            FractionAlongEdgeControl(
+                fraction = associationResult.getFractionAlongEdge()!!.toFloat(),
```
```suggestion
                fraction = fraction.toFloat(),
```

sorenoid commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/element/utilityassociationselement/UtilityAssociationsElementState.kt:
```diff
@@ -0,0 +1,119 @@
+    val description: String = element.description
+    val displayCount: Int = element.displayCount
+
+    private var _loading: MutableState<Boolean> = mutableStateOf(true)
+
+    private var _filters: MutableState<List<UtilityAssociationsFilterResult>> =
+        mutableStateOf(emptyList())
```
nitpick: normally a backing var is directly above the val that uses it.

sorenoid commented on toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/PopupState.kt:
```diff
@@ -0,0 +1,383 @@
+                }
+
+                else -> {
+                    // TODO remove for release
+                    println("encountered element of type ${element::class.java}")
```
don't forget to remove this!

sorenoid reviewed (changes requested):
Wow! that's quite an undertaking. I will test it out tomorrow. For now here is my review of the code.

puneet-pdx committed:
address code review comments

puneet-pdx committed:
FIx navigation issue

sorenoid reviewed (approved):
Great work!

puneet-pdx commented:
Thanks @kaushikrw and @sorenoid for the reviews! Merging this.
